File: src/main/java/algo/backtrack/CombinationSum_39.java
package algo.backtrack;

import java.util.*;

public class CombinationSum_39 {
    public static List<List<Integer>> combinationSum(int[] candidates, int target) {
        var combination = new ArrayList<Integer>();
        var comSum = new ArrayList<List<Integer>>();
        Arrays.sort(candidates);
        backtrack(candidates, target, 0, comSum, combination);
        return comSum;
    }

    private static void backtrack(int[] candidates, int target, int start, List<List<Integer>> result, List<Integer> combination) {
        if (target == 0) {
            result.add(new ArrayList<>(combination));
            return;
        }
        for (var i = start; i < candidates.length; i++) {
            if (candidates[i] <= target) {
                combination.add(candidates[i]);
                backtrack(candidates, target - candidates[i], i, result, combination);
                combination.removeLast();
            }
        }
    }

    public static void main(String[] args) {
        System.out.println(combinationSum(new int[]{2,3,6,7}, 7));
        System.out.println(combinationSum(new int[]{2,3,5}, 8));
        System.out.println(combinationSum(new int[]{2}, 1));
        System.out.println(combinationSum(new int[]{7,3,2}, 18));
    }
}

File: src/main/kotlin/algo/backtrack/CombinationSum_39.kt
package algo.backtrack

class KtCombinationSum_39 {
    fun combinationSum(candidates: IntArray, target: Int): List<List<Int>> {
        val combinations: MutableList<List<Int>> = mutableListOf()
        fun backtrack(candidates: IntArray, target: Int, list: MutableList<Int>) {
            candidates.forEach {
                val remain = target - it
                if (remain >= 0) {
                    list.add(it)
                    val sorted = list.sorted()
                    if (remain == 0) {
                        if (combinations.none { lst -> lst == sorted }) {
                            combinations.add(sorted)
                        }
                    } else {
                        backtrack(candidates, remain, list)
                    }
                    list.removeLast()
                }
            }
        }

        backtrack(candidates, target, mutableListOf())
        return combinations
    }

}

fun main() {
    val comSum = KtCombinationSum_39()
    println(comSum.combinationSum(intArrayOf(2,3,6,7), 7))
    println(comSum.combinationSum(intArrayOf(2,3,5), 8))
    println(comSum.combinationSum(intArrayOf(7,3,2), 18))
}